[John Ng] iP

JohnNub opened the pull request:
Create PR for iP

JohnNub committed:
Implement Level--0

JohnNub committed:
Implement level-1

JohnNub committed:
Implement Level-2
Add class Task

JohnNub committed:
Implement Level-3

Tidy up the main code with more helper functions to make it easy
to read and manage.
Implement automatic formatting for chatbot replies to ensure
uniformity.
toString implemented in Task as a shortcut for printing.

JohnNub committed:
Implement Level-4

Add ToDo, Deadline and Task classes.

JohnNub committed:
Implement A-TextUiTesting

homingjun commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,78 @@
+    public String toString() {
+        return "[D]" + super.toString() + " Due: " + due;
+    }
+}
```
Code naming is good and in accordance with java coding standards.

homingjun commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,117 @@
+            }
+        }
+        printOutput("Bye. Hope to see you again soon!", true);
+    }
```
The while loop is a little long and affects the ease of readability. Consider shortening the loop or separate some code into functions/classes.

homingjun reviewed (commented):
Overall good code naming and organisation, but some parts mentioned can be improved on.

Johnson-Yee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,117 @@
+                printOutput("Added new todo: " + t.toString());
+            } else {
+                printOutput("Commands: bye list todo deadline event", false);
+            }
```
The while loop could be too extensive and hard to read. Do consider changing the lines of code into methods such as commandDeadline. And on top of that, the while loop could be refactored into another method eg executeCommands. :D

Johnson-Yee commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,85 @@
+            } else if (dateStr.toLowerCase().matches("next\\s*sat.*")) {
+                this.start = LocalDateTime.now().with(next(DayOfWeek.SATURDAY));
+            } else if (dateStr.toLowerCase().matches("next\\s*sun.*")) {
+                this.start = LocalDateTime.now().with(next(DayOfWeek.SUNDAY));
```
Neat form of if-else statements! Good Job!

Johnson-Yee commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,34 @@
+        return isDone;
+    }
+
+    public void setDone(boolean done) {
```
Just a small change here: Boolean done should be named isDone

Johnson-Yee commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,78 @@
+import java.text.ParseException;
+import java.time.DayOfWeek;
+import java.time.LocalDate;
```
Good job on the explicit importing of the specific classes.

homingjun commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,117 @@
+                System.out.print(System.lineSeparator());
+            }
+        }
+        System.out.println("____________________________________________________________");
```
Try not to have nested loops (AKA arrow head) in your code. Try to refactor some parts of this loop to prevent this from happening.

homingjun reviewed (commented):
Overall good code quality and readability. However, some minor changes can be made to improve the code.

JohnNub committed:
Refactor code to tidy the structure.

JohnNub committed:
Add detailed exception handling for DukeException,
IndexOutOfBoundsException and NumberFormatException in main

JohnNub committed:
Refactor code into packages

JohnNub committed:
Merge branch 'branch-Level-5'

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+
+
+public class Duke {
+    static ArrayList<Task> taskList = new ArrayList<Task>();
```
I think you're missing an access modifier here. Should it be `public` or `private` etc?

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+        System.out.println("Hello from\n" + logo);
+        printOutput("Hello! I'm duke.main.Duke\n" +
+                "What can I do for you?", true);
+        while (true) {
```
I would suggest the use `switch` block here as you have a `break` in line 30.

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+            printOutput("Oops the list is empty!", false);
+            return;
+        }
+        String tl = "";
```
The variable `tl` here is not very informative. You might want to give it a more informative naming.

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+        Scanner in = new Scanner(System.in);
+
+        String userInput;
+        String logo = " ____        _        \n"
```
Although not necessary, the `logo` here can be considered a `constant` so you can further extract and refactor out.

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+
+    private static void setDone(String userInput) {
+        try {
+            String taskStr = userInput.substring(4).strip();
```
I have noticed you use `4` here and this seems like a magic number. Perhaps you can consider using an enumerator class here as well.

JosephLimWeiJie replied:
The same can be applied for all other `magic` numbers in your other methods too.

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+    private static void printOutput(String text, boolean instant) {
+        final String UNDERSCORES = "____________________________________________________________";
+        System.out.println(UNDERSCORES);
+        //Split text according to the lines to format
```
Take note. Javadoc comments should be `// Split text according to...... to format.`. Be careful of the space after the `//` and a `.` at the end of the sentence.

JosephLimWeiJie replied:
The same also applies to all other instances where you used a `//` comment.

JosephLimWeiJie commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,150 @@
+     * @param text string to be printed
+     * @param instant whether the string is printed instantly
+     */
+    private static void printOutput(String text, boolean instant) {
```
Boolean variable should be `isInstant`.

JosephLimWeiJie commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    /**
+     * tryReadDate - Attempts to convert the user supplied date string to a machine date
```
For this line of javadoc, you don't have to specify the function name. You can begin with `Attempts to convert..`

JosephLimWeiJie replied:
The same applies to all your other classes as I've noticed this is repeated.

JosephLimWeiJie commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,36 @@
+        return isDone;
+    }
+
+    public void setDone(boolean done) {
```
The boolean variable `done` should be named `isDone`. However, this conflicts with your already defined `isDone` within your `Task` class. Could there be a better way to rewrite this?

JosephLimWeiJie reviewed (commented):
Good Job overall. Your error handling blocks are quite comprehensive. In general, your code is pretty clean and easy to follow. 

Keep up the good work!

JohnNub committed:
Implement Level-6
Refactor code to address review comments
Add delete command
Add javadoc comments for Duke class

JohnNub committed:
Implement Level-7
Implement serializable objects

JohnNub committed:
Implement A-MoreOOP

Refactor Duke.java to introduce class
Parser, Storage, Ui, UserInput
This major refactor enables better code tidyness and parsing
consistency between commands.

JohnNub committed:
Update README.md

Accidentally deleted branch for level-6 in previous week

JohnNub committed:
Merge branch 'branch-Level-6'

JohnNub committed:
Update README.md

Created new branch as previous branch was accidentally deleted

JohnNub committed:
Merge branch 'branch-Level-7'

JohnNub committed:
Implement Level-8

Add ability to indicate end of day with "end" keyword

JohnNub committed:
Implement Level-9

Add find command to perform case insensitive search on list

JohnNub committed:
Merge pull request #1 from JohnNub/branch-Level-8

Implement Level-8

JohnNub committed:
Merge pull request #2 from JohnNub/branch-Level-9

Implement Level-9

JohnNub committed:
Implement A-JavaDoc

Correction in formatting and additional javadoc added

JohnNub committed:
Merge pull request #3 from JohnNub/branch-A-JavaDoc

Implement A-JavaDoc

JohnNub committed:
Implement A-UserGuide

Code corrections and presentation fixes for final release

JohnNub committed:
Update README.md
